src/wait.rs

```diff
@@ -0,0 +1,288 @@
+//! Sometimes it is necessary to wait for a browser to achieve a certain state. For example,
+//! navigating to a page may be take bit of time. And the time may vary between different
+//! environments and test runs. Static delays can work around this issue, but also prolong the
+//! test runs unnecessarily. Longer delays have less flaky tests, but even more unnecessary wait
+//! time.
+//!
+//! To wait as optimal as possible, you can use asynchronous wait operations, which periodically
+//! check for the expected state, re-try if necessary, but also fail after a certain time and still
+//! allow you to fail the test. Allow for longer grace periods, and only spending the time waiting
+//! when necessary.
+//!
+//! # Basic usage
+//!
+//! By default all wait operations will time-out after 30 seconds and will re-check every
+//! 250 milliseconds. You can configure this using the [`Wait::at_most`] and [`Wait::every`]
+//! methods.
+//!
+//! Once configured, you can start waiting on some condition by using the [`Wait::on`] method. It
+//! accepts any type implementing the [`WaitCondition`] trait. For example:
+//!
+//! ```no_run
+//! # use fantoccini::{ClientBuilder, Locator};
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), fantoccini::error::CmdError> {
+//! # #[cfg(all(feature = "native-tls", not(feature = "rustls-tls")))]
+//! # let mut client = ClientBuilder::native().connect("http://localhost:4444").await.expect("failed to connect to WebDriver");
+//! # #[cfg(feature = "rustls-tls")]
+//! # let mut client = ClientBuilder::rustls().connect("http://localhost:4444").await.expect("failed to connect to WebDriver");
+//! # #[cfg(all(not(feature = "native-tls"), not(feature = "rustls-tls")))]
+//! # let mut client: fantoccini::Client = unreachable!("no tls provider available");
+//! // -- snip wrapper code --
+//! let button = client.wait().on(Locator::Css(
+//!     r#"a.button-download[href="/learn/get-started"]"#,
+//! )).await?;
+//! // -- snip wrapper code --
+//! # client.close().await
+//! # }
+//! ```
+//!
+//! # Error handling
+//!
+//! When a wait operation times out, it will return a [`CmdError::WaitTimeout`]. When a wait
+//! condition check returns an error, the wait operation will be aborted, and the error returned.
+//!
+//! # Custom conditions
+//!
+//! You can implement custom conditions either by implementing the [`WaitCondition`] trait or by
+//! using closures. Due to lifetime and async trait difficulties, two newtypes and two dedicated
+//! functions exists to simplify the usage of closures. Also see: [`Closure`], [`Predicate`].
```

Comment:
```suggestion
//! using closures. If you just want a closure, use [`Wait::until`] and [`Wait::until_some`].
```

I don't think anyone will want to use `Closure` and `Predicate` directly.